include/libtorrent/aux_/session_udp_sockets.hpp

```diff
@@ -56,6 +59,42 @@ namespace libtorrent { namespace aux {
 		bool write_blocked = false;
 	};
 
+	struct outgoing_udp_socket final : session_udp_socket
+	{
+		explicit outgoing_udp_socket(io_service& ios, std::string const& dev, bool ssl_)
+			: session_udp_socket(ios), device(dev), ssl(ssl_) {}
+
+		// the name of the device the socket is bound to, may be empty
+		// if the socket is not bound to a device
+		std::string device;
+
+		// set to true if this is an SSL socket
+		bool ssl = false;
+	};
+
+	// sockets used for outoing utp connections
+	struct TORRENT_EXTRA_EXPORT outgoing_sockets
+	{
+		// partitions sockets based on whether they match one of the given endpoints
+		// all matched sockets are ordered before unmatched sockets
+		// matched endpoints are removed from the vector
+		// returns an iterator to the first unmatched socket
+		std::vector<std::shared_ptr<outgoing_udp_socket>>::iterator
+		partition_outgoing_sockets(std::vector<listen_endpoint_t>& eps);
+
+		tcp::endpoint bind(socket_type& s, address const& remote_address) const;
+
+		void update_proxy(proxy_settings const& settings);
+
+		// close all sockets
+		void close();
+
+		std::vector<std::shared_ptr<outgoing_udp_socket>> sockets;
```

Comment:
`sockets` is accessed directly by `session_impl::reopen_outgoing_sockets`, index can be private.